File: Kruskal/feb2019haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph 

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal a = aux cola' dic' [] 
    where
        cola' = foldr (Q.enqueue) (Q.empty) (edges a) 
        dic'  =   foldr (\v d ->D.insert v v d ) (D.empty) (vertices a)
        
        aux cola dic lista
            |Q.isEmpty cola = lista
            |r1 == r2  = aux (Q.dequeue cola) dic lista
            |otherwise = aux (Q.dequeue cola) (D.insert r2 v1 dic) (we:lista)
            where
                r1 = representante v1 dic
                r2 = representante v2 dic
                we@(WE v1 peso v2) = Q.first cola


representante :: (Ord a) => a->D.Dictionary a a -> a
representante v dic
                |w==v = v
                |otherwise = representante w dic 
            where 
                Just w = D.valueOf v dic


-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined

File: Kruskal/feb2019java/feb2019java/src/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:Eduard Cybulkiewicz
 * -- GRADO/STUDIES: Inf A
 * -- NÚM. MÁQUINA/MACHINE NUMBER: 23
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.DictionaryWeightedGraph;
import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

public class Kruskal {
	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {

		Set<WeightedEdge<V,W>> set = new HashSet<>();

		Dictionary<V,V> d = new HashDictionary<>();

		for(V v : g.vertices()){
			d.insert(v,v);
		}

		LinkedPriorityQueue<WeightedEdge<V,W>> cola = new LinkedPriorityQueue<>();

		for(WeightedEdge<V, W> aristas : g.edges()){
			cola.enqueue(aristas);
		}
		while(!cola.isEmpty()){
			WeightedEdge<V,W> arista = cola.first();
			cola.dequeue();
			if(!(representante(arista.source(), d).equals(representante(arista.destination(), d)))){
				d.insert(representante(arista.destination(),d), arista.source());
				set.insert(arista);
			}
		}
		return set;
	}

	static <V> V representante(V src, Dictionary<V,V> dic ){

		if(src.equals(dic.valueOf(src))){
			return src;
		}else{
			return representante(dic.valueOf(src),dic);
		}
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}
